src/Compilation.zig

```diff
@@ -879,25 +880,23 @@ pub fn create(gpa: *Allocator, options: InitOptions) !*Compilation {
             break :blk false;
         };
 
-        const DarwinOptions = struct {
-            syslibroot: ?[]const u8 = null,
-            system_linker_hack: bool = false,
-        };
+        var darwin_system_linker_hack = false;
 
-        const darwin_options: DarwinOptions = if (build_options.have_llvm and comptime std.Target.current.isDarwin()) outer: {
-            const opts: DarwinOptions = if (use_lld and std.builtin.os.tag == .macos and options.target.isDarwin()) inner: {
+        const sysroot = blk: {
+            if (options.sysroot) |sysroot| {
+                break :blk sysroot;
+            } else if (build_options.have_llvm and comptime std.Target.current.isDarwin() and
```

Comment:
You'll need to either split the check into two so that `build_options.have_llvm and comptime std.Target.current.isDarwin()` is evaluated in compile-time, or remove the `comptime` altogether. `use_lld` and `options.target.isDarwin()` are only runtime-known.

Reply:
Ah I see. On looking at this again I realize that as an unintended side-effect of my changes the system linker hack will not work if `--sysroot` is passed. I'll look for a way to fix both of these things without duplicating logic.

Reply:
Actually, given that we are close to removing system linker hack altogether, I'm fine with not having it on with explicit `--sysroot`. Thoughts? @andrewrk care to weigh in?

Reply:
There is actually one major feature that's still missing, creation of dylibs, but apart from that, `zig ld` is more than capable.

Reply:
I've force-pushed again and revised this code to make the system linker hack compatible with `--sysroot` and hopefully fix the compile error.